src/backend/sgx/attestation.rs

```diff
@@ -0,0 +1,317 @@
+            format!(
+                "GeSupportedAttKeyIDs: invalid count: {} != 1",
+                key_ids.len()
+            ),
+        ));
+    }
+
+    Ok(key_ids.get(0).unwrap().clone())
+}
+
+/// Fills the Target Info of the QE into the output buffer specified and
+/// returns the number of bytes written.
+fn get_target_info(akid: Vec<u8>, size: usize, out_buf: &mut [u8]) -> Result<usize, Error> {
+    if out_buf.len() != SGX_TI_SIZE {
+        return Err(Error::new(
+            ErrorKind::Other,
+            format!(
+                "Invalid output buffer size: {} != {}",
+                out_buf.len(),
+                SGX_TI_SIZE
+            ),
+        ));
+    }
+
+    let mut req = AesmRequest::new();
+    let mut msg = Request_InitQuoteExRequest::new();
+
+    msg.set_timeout(AESM_REQUEST_TIMEOUT);
+    msg.set_b_pub_key_id(true);
+    msg.set_att_key_id(akid);
+    msg.set_buf_size(size as u64);
+    req.set_initQuoteExReq(msg);
+
+    let pb_msg = req.send()?;
+
+    let res = pb_msg.get_initQuoteExRes();
+
+    if res.get_errorCode() != 0 {
+        return Err(Error::new(
+            ErrorKind::InvalidData,
+            format!("InitQuoteExRequest: error: {:?}", res.get_errorCode()),
+        ));
+    }
+
+    let ti = res.get_target_info();
+
+    if ti.len() != SGX_TI_SIZE {
+        return Err(Error::new(
+            ErrorKind::Other,
+            format!(
+                "InitQuoteEx: Invalid TARGETINFO size: {} != {}",
+                ti.len(),
+                SGX_TI_SIZE
+            ),
+        ));
+    }
+
+    out_buf.copy_from_slice(ti);
+
+    Ok(ti.len())
+}
+
+/// Gets key size
+fn get_key_size(akid: Vec<u8>) -> Result<usize, Error> {
+    let mut req = AesmRequest::new();
+    let mut msg = Request_InitQuoteExRequest::new();
+
+    msg.set_timeout(AESM_REQUEST_TIMEOUT);
+    msg.set_b_pub_key_id(false);
+    msg.set_att_key_id(akid);
+    req.set_initQuoteExReq(msg);
+
+    let pb_msg = req.send()?;
+
+    let res = pb_msg.get_initQuoteExRes();
+
+    if res.get_errorCode() != 0 {
+        return Err(Error::new(
+            ErrorKind::InvalidData,
+            format!("InitQuoteEx error: {:?}", res.get_errorCode()),
+        ));
+    }
+
+    Ok(res.get_pub_key_id_size() as usize)
+}
+
+/// Fills the Quote obtained from the AESMD for the Report specified into
+/// the output buffer specified and returns the number of bytes written.
+fn get_quote(report: &[u8], akid: Vec<u8>, out_buf: &mut [u8]) -> Result<usize, Error> {
+    if out_buf.len() != SGX_QUOTE_SIZE {
+        return Err(Error::new(
+            ErrorKind::InvalidData,
+            format!(
+                "Invalid size of output buffer {} != {}",
+                out_buf.len(),
+                SGX_QUOTE_SIZE
+            ),
+        ));
+    }
+
+    let mut req = AesmRequest::new();
+
+    let mut msg = Request_GetQuoteExRequest::new();
+    msg.set_timeout(AESM_REQUEST_TIMEOUT);
+    msg.set_report(report[0..432].to_vec());
```

Comment:
Do we want to extract this (`432`) into a constant?

Reply:
```
pub struct Report {
    pub payload: ReportPayload,
    pub keyid: [u8; 32],
    pub mac: [u64; 2],
    padding: [u8; 80],
}
```
So, the `432` is the size of `Report` minus the padding bytes ... `\o/`

Reply:
At the very least there should be a comment here

Reply:
agreed

Reply:
I added `SGX_REPORT_SIZE` to `attestation.rs`.